cache/sharded_cache.h

```diff
@@ -9,29 +9,36 @@
 
 #pragma once
 
+#include <algorithm>
 #include <atomic>
+#include <cstdint>
+#include <memory>
 #include <string>
 
 #include "port/port.h"
 #include "rocksdb/cache.h"
+#include "util/hash.h"
+#include "util/math.h"
+#include "util/mutexlock.h"
 
 namespace ROCKSDB_NAMESPACE {
 
 // Single cache shard interface.
+template <typename HashType>
 class CacheShard {
  public:
   CacheShard() = default;
   virtual ~CacheShard() = default;
 
   using DeleterFn = Cache::DeleterFn;
-  virtual Status Insert(const Slice& key, uint32_t hash, void* value,
+  virtual Status Insert(const Slice& key, const HashType& hash, void* value,
```

Comment:
We use const references for hashes, which can be complex types.

Reply:
Depending on function inlining, this might cause performance regression when HashType is uint32_t. Needs performance testing, preferably with both gcc (default) and clang (USE_CLANG=1).